tools/chplcheck/src/rules.py

```diff
@@ -95,42 +101,88 @@ def MethodsAfterFields(context, node):
-            return None
-
-        def recurse(node, skip_direct = False):
+                var_node = node
+            
+            if var_node is None:
+                return None
+            
+            var_type = None
+            var_type_expr = var_node.type_expression()
+
+            if isinstance(var_type_expr, FnCall):
+                #for function call, we need to match all the components
+                var_type = ''
+                for child in var_type_expr:
+                    if child is None:
+                        continue
+                    if 'name' in dir(child):
+                        var_type += child.name()
+                    elif 'text' in dir(child):
+                        var_type += child.text()
+            elif isinstance(var_type_expr, Identifier):
+                var_type = var_type_expr.name()
+            
+            var_kind = var_node.kind()
+
+            var_attributes = ''
+            var_attribute_group = var_node.attribute_group()
+            if var_attribute_group is not None:
+                for attribute_node in var_attribute_group:
+                    if attribute_node is not None:
+                        var_attributes += attribute_node.name() + " "
+            
+            var_linkage = var_node.linkage()
+
+            var_pragmas = ' '.join(var_node.pragmas())
+            return (var_type, var_kind, var_attributes, var_linkage, var_pragmas)
+        
+        def recurse(node):
             consecutive = []
-            last_kind = None
-            last_has_attribute = False
+            last_characteristics = None
 
+            
             for child in node:
-                yield from recurse(child, skip_direct = isinstance(child, MultiDecl))
-
-                if skip_direct: continue
-
-                new_kind = is_relevant_decl(child)
-                has_attribute = child.attribute_group() is not None
-                any_has_attribute = last_has_attribute or has_attribute
-                compatible_kinds = not any_has_attribute and (last_kind is None or last_kind == new_kind)
-                last_kind = new_kind
-                last_has_attribute = has_attribute
-
-                # If we ran out of compatible decls, see if we can return them.
-                if not compatible_kinds:
+                #we want to skip Comments entirely
+                if isinstance(child,Comment):
+                    continue
+                
+                #we want to do MultiDecls and TupleDecls, but not recurse
+                skip_children = isinstance(child, MultiDecl) \
+                    or isinstance(child, TupleDecl) \
+                
+                if not skip_children:
+                    yield from recurse(child)
+
+                new_characteristics = is_relevant_decl(child)
+                compatible = new_characteristics is not None and \
+                    new_characteristics == last_characteristics
+                
```

Comment:
The check for None is not needed

Reply:
I think it is; `is_relavent_decl` returns None for non-`Variable` nodes on line 121